File: swordOffer/maxSubArray/maxSubArray.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  int maxSubArray(vector<int>& nums) {
    if (nums.empty()) {
      return 0;
    }

    vector<int> dp;
    dp.reserve(nums.size());
    int res = 0;
    for (int i = 0; i < nums.size(); i++) {
      if (i - 1 < 0) {
        dp[i] = nums[i];
        res = dp[i];
        continue;
      }
      dp[i] = max(nums[i], dp[i - 1] + nums[i]);
      if (dp[i] > res) {
        res = dp[i];
      }
    }

    return res;
  }
};

int main() {
  auto s = new Solution;
  vector<int> input = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
  cout << s->maxSubArray(input) << endl;
}

File: swordOffer/go/dp/maxSubArray.go
package main

func maxSubArray(nums []int) int {
	dp := make([]int, len(nums))
	dp[0] = nums[0]
	res := nums[0]
	for i := 1; i < len(nums); i++ {
		if dp[i-1] > 0 {
			dp[i] = dp[i-1] + nums[i]
		} else {
			dp[i] = nums[i]
		}
		res = max(res, dp[i])
	}

	return res
}

/*
O(n2) solution with prefix sum
func maxSubArray(nums []int) int {
	pre := make([]int, len(nums))
	pre[0] = nums[0]
	for i := 1; i < len(nums); i++ {
		pre[i] = pre[i-1] + nums[i]
	}

	res := nums[0]
	for i := 0; i < len(nums); i++ {
		res = max(res, pre[i])
		for j := i + 1; j < len(nums); j++ {
			sum := pre[j] - pre[i]
			res = max(res, sum)
		}
	}

	return res
}
*/

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
